### PR TITLE
BUG: pd.read_json May Not Maintain Numeric String Index

### LINKED ISSUE #28556: pd.read_json May Not Maintain Numeric String Index
```python
>>> df = pd.DataFrame(range(3), index=list("123"))
>>> df.to_json(orient="split")
'{"columns":[0],"index":["1","2","3"],"data":[[0],[1],[2]]}'
>>> pd.read_json(df.to_json(orient="split"), orient="split").index
Int64Index([1, 2, 3], dtype='int64')
```

Note that the string nature of the values should be preserved via roundtrip here, but ends up being lossy anyway. Noted during refactor of #28510

### PR DESCRIPTION
**Pull request by theoniko**
- [x] closes #28556
- [x] tests added / passed
- [ ] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/json/_json.py [hunk 1]
```diff
@@ -850,6 +850,26 @@ def _convert_axes(self):
         obj = self.obj
         assert obj is not None  # for mypy
         for axis_name in obj._AXIS_ORDERS:
+            if (
+                self.orient == "split"
+                and axis_name == "index"
+                and obj._get_axis(axis_name).dtype == "object"
+                and obj.index.inferred_type == "string"
+            ):
+                index_list = list(obj._get_axis(axis_name).values)
```
Comment (arw2019) [inline pandas/io/json/_json.py:859]: do we have to convert to a list here? We should be able to inspect `values` directly I think

File: pandas/io/json/_json.py [hunk 2]
```diff
@@ -850,6 +850,26 @@ def _convert_axes(self):
...
+                are_all_strings = all(type(element) == str for element in index_list)
```
Comment (arw2019) [inline pandas/io/json/_json.py:860]: we have helpers got these kinds of checks in `pandas/core/dtypes/common.py`, may as well use those

File: pandas/io/json/_json.py [hunk 3]
```diff
@@ -850,6 +850,26 @@ def _convert_axes(self):
...
+                are_int_or_floats = [
+                    True
+                    for element in index_list
+                    if isinstance(element, int) or isinstance(element, float)
+                ]
+                are_all_int_or_floats = all(
+                    are_int_or_floats is True for element in are_int_or_floats
+                )
+                if are_all_int_or_floats and are_all_strings:
```
Comment (arw2019) [inline pandas/io/json/_json.py:869]: is this ever hit? only one of those can be true at any time, no?

File: pandas/io/json/_json.py [hunk 4]
```diff
@@ -913,7 +933,6 @@ def _try_convert_data(self, name, data, use_dtypes=True, convert_dates=True):
                     result = True
                 except (TypeError, ValueError):
                     pass
-
```
Comment (arw2019) [inline pandas/io/json/_json.py:916]: pls revert this to minimize the diff

File: pandas/tests/io/json/test_pandas.py [hunk 5]
```diff
@@ -192,11 +192,14 @@ def test_roundtrip_str_axes(self, orient, convert_axes, numpy, dtype):
         # to disambiguate whether those keys actually were strings or numeric
         # beforehand and numeric wins out.
         # TODO: Split should be able to support this
```
Comment (arw2019) [inline pandas/tests/io/json/test_pandas.py:194]: can we delete this TODO now?

**Review by arw2019 — CHANGES_REQUESTED** [review]
Thanks @theoniko for the PR!

Some comments. We'll also need a whatsnew for this (1.3)

File: pandas/io/json/_json.py [hunk 6]
```diff
@@ -850,6 +850,26 @@ def _convert_axes(self):
         obj = self.obj
         assert obj is not None  # for mypy
         for axis_name in obj._AXIS_ORDERS:
+            if (
```
Comment (mroeschke) [inline pandas/io/json/_json.py:853]: This `if` condition seems too specific. Is there a more generic way to handle this?

**Review by jreback — CHANGES_REQUESTED** [review]
needs to be less specialized (as indicated by comments)

File: pandas/io/json/_json.py (hunk 6 above)
Comment (theoniko) [reply pandas/io/json/_json.py:853]: As far as i understood the minimum preconditions to keep string index should in read_json  orient equal to split or index and the index of json format to be a list of int or float strings. Am i right? Should i move the checks in _try_convert_data?
Do you have a more conrete and robust suggestion?

Comment (theoniko) [reply pandas/io/json/_json.py]: I tried a slightly different way. Could you please re-review?

File: pandas/tests/io/json/test_pandas.py [hunk 7]
```diff
@@ -192,11 +192,14 @@ def test_roundtrip_str_axes(self, orient, convert_axes, numpy, dtype):
...
-        if convert_axes and (orient in ("split", "index", "columns")):
+        if convert_axes and (orient in ("index", "columns")):
             expected.columns = expected.columns.astype(np.int64)
             expected.index = expected.index.astype(np.int64)
         elif orient == "records" and convert_axes:
             expected.columns = expected.columns.astype(np.int64)
+        elif convert_axes and orient == "split":
+            expected.columns = expected.columns.astype(np.int64)
+            expected.index = expected.index.astype(str)
```
Comment (WillAyd) [inline pandas/tests/io/json/test_pandas.py:202]: Do we need this conversion? I think the problem is that we are converting when we don't need to

**Comment by theoniko** [comment]
All requested changes were made. Please re-review the pr.

File: pandas/io/json/_json.py [hunk 8]
```diff
@@ -894,14 +895,19 @@ def _try_convert_data(self, name, data, use_dtypes=True, convert_dates=True):
             if result:
                 return new_data, True
 
-        result = False
+        result = True
 
         if data.dtype == "object":
-
+            if (
+                isinstance(data, Index)
```
Comment (jreback) [inline pandas/io/json/_json.py:902]: i don't think you need to test if an Index

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
merge master as well

File: doc/source/whatsnew/v1.3.0.rst [hunk 9]
```diff
@@ -267,7 +267,7 @@ I/O
 - Bug in :func:`to_hdf` raising ``KeyError`` when trying to apply
   for subclasses of ``DataFrame`` or ``Series`` (:issue:`33748`).
 - Bug in :func:`json_normalize` resulting in the first element of a generator object not being included in the returned ``DataFrame`` (:issue:`35923`)
-
+- Bug in :func:`read_json` does not maintan numeric string index (:issue:`28556`)
```
[the remark this answers is not in the document]
Comment (theoniko) [reply doc/source/whatsnew/v1.3.0.rst]: Done

File: pandas/io/json/_json.py [hunk 10]
```diff
@@ -894,14 +895,19 @@ def _try_convert_data(self, name, data, use_dtypes=True, convert_dates=True):
...
+                and name == "index"
+                and self.orient == "split"
+                and len(data)
+            ):
+                result = False
```
[the remark this answers is not in the document]
Comment (theoniko) [reply pandas/io/json/_json.py]: I moved the condition to the end and removed the result variable.

File: pandas/io/json/_json.py (hunk 8 above)
Comment (theoniko) [reply pandas/io/json/_json.py]: Done

File: pandas/io/json/_json.py [hunk 11]
```diff
@@ -934,11 +929,13 @@ def _try_convert_data(self, name, data, use_dtypes=True, convert_dates=True):
             # coerce floats to 64
             try:
                 data = data.astype("int64")
-                result = True
             except (TypeError, ValueError):
                 pass
 
-        return data, result
+        if name == "index" and self.orient == "split" and len(data):
```
Comment (jreback) [inline pandas/io/json/_json.py:935]: ok looks a lot better. can you add a comment here and structure this more like

```
# if we have an index, we want to preserve dtypes
if name == "index" and len(data):
    if self.orient == "split":
        return data, False

return data, True
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/json/_json.py (hunk 11 above)
Comment (theoniko) [reply pandas/io/json/_json.py]: Done

File: doc/source/whatsnew/v1.3.0.rst [hunk 12]
```diff
@@ -272,6 +272,7 @@ I/O
 - Bug in :func:`json_normalize` resulting in the first element of a generator object not being included in the returned ``DataFrame`` (:issue:`35923`)
 - Bug in :func:`read_excel` forward filling :class:`MultiIndex` names with multiple header and index columns specified (:issue:`34673`)
 - :func:`pandas.read_excel` now respects :func:``pandas.set_option`` (:issue:`34252`)
+- Bug in :func:`read_json` when ``orient="split"`` does not maintan numeric string index (:issue:`28556`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:275]: Bug in :func:`read_json` when ``orient="split"`` does not maintain numeric string index (:issue:`28556`)

Comment (theoniko) [reply doc/source/whatsnew/v1.3.0.rst]: Done.

**Review by jreback — APPROVED** [review]

File: doc/source/whatsnew/v1.3.0.rst [hunk 13]
```diff
@@ -274,6 +274,7 @@ I/O
 - Bug in :func:`json_normalize` resulting in the first element of a generator object not being included in the returned ``DataFrame`` (:issue:`35923`)
 - Bug in :func:`read_excel` forward filling :class:`MultiIndex` names with multiple header and index columns specified (:issue:`34673`)
 - :func:`pandas.read_excel` now respects :func:``pandas.set_option`` (:issue:`34252`)
+- Bug in :func:``read_json`` when ``orient="split"`` does not maintan numeric string index (:issue:``28556``)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:277]:
```suggestion
- Bug in :func:``read_json`` when ``orient="split"`` does not maintan numeric string index (:issue:`28556`)
```

**Comment by jreback** [comment]
lgtm pushed a small doc change and can merge on green

**Pull request merged**